services/logs/filter.go

```diff
@@ -0,0 +1,216 @@
+	id common.Hash,
+	criteria FilterCriteria,
+	blocks storage.BlockIndexer,
+	receipts storage.ReceiptIndexer,
+) *IDFilter {
+	return &IDFilter{
+		id:       id,
+		criteria: criteria,
+		blocks:   blocks,
+		receipts: receipts,
+	}
+}
+
+func (i *IDFilter) Match() ([]*gethTypes.Log, error) {
+	blk, err := i.blocks.GetByID(i.id)
+	if err != nil {
+		return nil, err
+	}
+
+	receipt, err := i.receipts.GetByBlockHeight(big.NewInt(int64(blk.Height)))
+	if err != nil {
+		return nil, err
+	}
+
+	logs := make([]*gethTypes.Log, 0)
+	for _, log := range receipt.Logs {
+		if exactMatch(log, i.criteria) {
+			logs = append(logs, log)
+		}
+	}
+
+	return logs, nil
+}
+
+// StreamFilter matches all the logs against the criteria from the receipt channel.
+type StreamFilter struct {
+	criteria      FilterCriteria
+	receiptStream chan *gethTypes.Receipt
+}
+
+func NewStreamFilter(criteria FilterCriteria, receipts chan *gethTypes.Receipt) *StreamFilter {
+	return &StreamFilter{
+		criteria:      criteria,
+		receiptStream: receipts,
+	}
+}
+
+func (s *StreamFilter) Match() (chan *gethTypes.Log, error) {
+	logs := make(chan *gethTypes.Log)
+
+	go func() {
+		defer close(logs)
+
+		for {
+			select {
+			case receipt, ok := <-s.receiptStream:
+				if !ok {
+					return // exit the goroutine if receiptStream is closed
+				}
+
+				if !bloomMatch(receipt.Bloom, s.criteria) {
+					continue
+				}
+
+				for _, log := range receipt.Logs {
+					if exactMatch(log, s.criteria) {
+						logs <- log
+					}
+				}
+			}
+		}
+	}()
+
+	return logs, nil
+}
+
+// exactMatch checks the topic and address values of the log match the filer exactly.
+func exactMatch(log *gethTypes.Log, criteria FilterCriteria) bool {
+	if len(criteria.Topics) > len(log.Topics) {
+		return false
+	}
+
+	for _, sub := range criteria.Topics {
+		for _, topic := range sub {
+			if !slices.Contains(log.Topics, topic) {
+				return false
+			}
+		}
+	}
+
+	if !slices.Contains(criteria.Addresses, log.Address) {
+		return false
+	}
+
+	return true
+}
+
+// bloomMatch takes a bloom value and tests if the addresses and topics provided pass the bloom filter.
+// This acts as a fast probabilistic test that might produce false-positives but not false-negatives.
+// If true is returned we should further check against the exactMatch to really make sure the log is matched.
+//
+// source: https://github.com/ethereum/go-ethereum/blob/8d1db1601d3a9e4fd067558a49db6f0b879c9b48/eth/filters/filter.go#L395
+func bloomMatch(bloom gethTypes.Bloom, criteria FilterCriteria) bool {
+	if len(criteria.Addresses) > 0 {
+		var included bool
+		for _, addr := range criteria.Addresses {
+			if gethTypes.BloomLookup(bloom, addr) {
+				included = true
+				break
+			}
+		}
+		if !included {
+			return false
+		}
+	}
+
+	for _, sub := range criteria.Topics {
+		included := len(sub) == 0 // empty rule set == wildcard
+		for _, topic := range sub {
+			if gethTypes.BloomLookup(bloom, topic) {
+				included = true
```

Comment:
Same here, will an early return work in this case?

Reply:
again no, because we go over multiple topics and all have to true, not just the first

Reply:
Ah, I see, it's not meant for `or` match :+1: